refactor: optimize block index comparisons (1.4-6.8x faster)

l0rinc opened the pull request:
**Context:** The comparator is often called either directly or for the sorting of [`setBlockIndexCandidates`](https://github.com/bitcoin/bitcoin/blob/master/src/validation.h#L637), a red-black tree-set containing valid block headers where the comparator is invoked extensively.
Besides the sorted set usages, the comparator is used directly in `Chainstate` in `FindMostWorkChain`, `PruneBlockIndexCandidates`, `ActivateBestChain`, and `InvalidateBlock`; and in `ChainstateManager` in `LoadBlockIndex`, `CheckBlockIndex`, `ActivateSnapshot`, and `PopulateAndValidateSnapshot`.

It was profiled during the investigation of https://github.com/bitcoin/bitcoin/issues/33618#issuecomment-3402229976.

**Testing:** To ensure the optimized implementations are both fast and correct, the first commit adds a dedicated benchmark to measure `CBlockIndexWorkComparator` performance, and the second commit adds deterministic sorting tests comparing the new implementation with the original one. Later commits add fuzz targets for equivalence coverage.

**Results:**
> GCC 15.0.1:

* `CBlockIndexWorkComparator`: 100,772,511 cmp/s → 656,674,205 cmp/s = 6.51x faster
* `CheckBlockIndex`: 9,091 ops/s → 14,765 ops/s = 1.62x faster
> Clang 22.0.0:

* `CBlockIndexWorkComparator`: 100,451,893 cmp/s → 683,414,234 cmp/s = 6.8x faster
* `CheckBlockIndex`: 10,322 ops/s → 14,376 ops/s = 1.39x faster

See https://corecheck.dev/bitcoin/bitcoin/pulls/33637 for the CI's `CheckBlockIndex` performance.
<details>
<summary>gcc and clang measurements</summary>

> Compiler: gcc 15.0.1

b60450fae8 bench: add benchmark to measure `CBlockIndexWorkComparator` performance

|              ns/cmp |               cmp/s |    err% |         ins/cmp |         cyc/cmp |    IPC |        bra/cmp |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|                9.92 |      100,772,511.62 |    0.0% |           63.98 |           35.64 |  1.795 |          14.17 |    1.9% |      5.50 | `CBlockIndexWorkComparator`

|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|          109,996.46 |            9,091.20 |    0.2% |    1,014,421.11 |      394,979.29 |  2.568 |     313,025.11 |    0.0% |      5.50 | `CheckBlockIndex`

e2e02177ba refactor: inline `arith_uint256` comparison operator

|              ns/cmp |               cmp/s |    err% |         ins/cmp |         cyc/cmp |    IPC |        bra/cmp |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|                6.48 |      154,439,491.66 |    0.0% |           31.01 |           23.25 |  1.334 |           7.16 |    3.8% |      5.50 | `CBlockIndexWorkComparator`

|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|          105,754.86 |            9,455.83 |    0.1% |      913,130.11 |      379,588.20 |  2.406 |     276,692.11 |    0.0% |      5.50 | `CheckBlockIndex`

85b74b01de refactor: optimize `arith_uint256` comparison with spaceship operator

|              ns/cmp |               cmp/s |    err% |         ins/cmp |         cyc/cmp |    IPC |        bra/cmp |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|                6.37 |      156,990,488.06 |    0.0% |           28.85 |           22.87 |  1.261 |           8.61 |    3.2% |      5.50 | `CBlockIndexWorkComparator`

|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|           83,803.10 |           11,932.73 |    0.0% |      743,565.09 |      300,824.84 |  2.472 |     232,646.08 |    0.0% |      5.56 | `CheckBlockIndex`

deb58eea2f refactor: optimize `CBlockIndexWorkComparator` with std::tie

|              ns/cmp |               cmp/s |    err% |         ins/cmp |         cyc/cmp |    IPC |        bra/cmp |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|                1.52 |      656,674,205.98 |    0.0% |           13.18 |            5.47 |  2.410 |           3.06 |    0.1% |      5.50 | `CBlockIndexWorkComparator`

|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|           67,726.84 |           14,765.19 |    0.2% |      585,826.07 |      243,155.90 |  2.409 |     181,920.07 |    0.0% |      5.54 | `CheckBlockIndex`
> Compiler: clang 22.0.0

b60450fae8 bench: add benchmark to measure `CBlockIndexWorkComparator` performance

|              ns/cmp |               cmp/s |    err% |         ins/cmp |         cyc/cmp |    IPC |        bra/cmp |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|                9.96 |      100,451,893.46 |    0.0% |           61.28 |           35.75 |  1.714 |          13.62 |    2.1% |      5.52 | `CBlockIndexWorkComparator`

|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|           96,878.70 |           10,322.19 |    0.1% |      802,827.10 |      347,679.01 |  2.309 |     234,823.10 |    0.0% |      5.50 | `CheckBlockIndex`

e2e02177ba refactor: inline `arith_uint256` comparison operator

|              ns/cmp |               cmp/s |    err% |         ins/cmp |         cyc/cmp |    IPC |        bra/cmp |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|                6.13 |      163,183,693.10 |    0.0% |           25.74 |           22.01 |  1.170 |           6.10 |    4.6% |      5.50 | `CBlockIndexWorkComparator`

|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|           86,307.94 |           11,586.42 |    1.2% |      646,229.09 |      309,785.33 |  2.086 |     195,119.09 |    0.0% |      5.54 | `CheckBlockIndex`

85b74b01de refactor: optimize `arith_uint256` comparison with spaceship operator

|              ns/cmp |               cmp/s |    err% |         ins/cmp |         cyc/cmp |    IPC |        bra/cmp |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|                6.36 |      157,330,900.16 |    1.0% |           26.20 |           22.61 |  1.159 |           6.55 |    4.4% |      5.53 | `CBlockIndexWorkComparator`

|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|           75,031.66 |           13,327.71 |    1.0% |      650,604.15 |      266,934.04 |  2.437 |     149,967.14 |    0.0% |      5.54 | `CheckBlockIndex`

deb58eea2f refactor: optimize `CBlockIndexWorkComparator` with std::tie

|              ns/cmp |               cmp/s |    err% |         ins/cmp |         cyc/cmp |    IPC |        bra/cmp |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|                1.46 |      683,414,234.14 |    0.0% |           16.12 |            5.25 |  3.067 |           4.02 |    0.1% |      5.50 | `CBlockIndexWorkComparator`

|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|           69,559.70 |           14,376.14 |    0.1% |      559,208.07 |      249,654.28 |  2.240 |     132,342.07 |    0.0% |      5.51 | `CheckBlockIndex`
</details>
**Reproducer:** Run the equivalence tests with:
```bash
cmake -B build && cmake --build build && ./build/bin/test_bitcoin --run_test=arith_uint256_tests,blockchain_tests
```
Each commit shows how it changes the relevant benchmarks.
<details>
<summary>Benchmark script</summary>

```bash
for compiler in gcc clang; do \
  if [ "$compiler" = "gcc" ]; then CC=gcc; CXX=g++; COMP_VER=$(gcc -dumpfullversion); \
  else CC=clang; CXX=clang++; COMP_VER=$(clang -dumpversion); fi && \
  echo "> Compiler: $compiler $COMP_VER" && \
  for commit in 70c6244b709fcf6a853ea44b91b9ed219609ac90 \
                516e05cd7d8dbb61b05854c59f5393367e502284 \
                d053b27c06d06cccbc6395b0aa7dc1c4193a6fc3 \
                8ef4ce2cb3ac80cbe58b8cf76e183f9eca6fafe4 \
                703165e6d3b9fbd58b9c31ddce0e961047fc5e71; do \
    git fetch origin $commit >/dev/null 2>&1 && git checkout $commit >/dev/null 2>&1 && git log -1 --pretty='%h %s' && \
    rm -rf build && \
    cmake -B build -DBUILD_BENCH=ON -DENABLE_IPC=OFF -DCMAKE_BUILD_TYPE=RelWithDebInfo \
      -DCMAKE_C_COMPILER=$CC -DCMAKE_CXX_COMPILER=$CXX >/dev/null 2>&1 && \
    cmake --build build -j$(nproc) >/dev/null 2>&1 && \
    for i in 1; do \
      build/bin/bench_bitcoin -filter='CBlockIndexWorkComparator|CheckBlockIndex' -min-time=5000; \
    done; \
  done; \
done
```
</details>
Note: something similar was already started in #33334, but this is a broader optimization that doesn't use the same technique.

l0rinc committed:
bench: add benchmark to measure `CBlockIndexWorkComparator` performance

Profiling shows this comparator consumes a significant portion of `CheckBlockIndex()`:
... ChainstateManager::CheckBlockIndex()
    ... std::_Rb_tree<...>::find(...)
        ... node::CBlockIndexWorkComparator::operator()(...)
            ... base_uint<256u>::CompareTo(...) const

This commit introduces a benchmark that performs pairwise comparisons on 1000 randomly generated block indices (with some duplicates) to establish baseline performance metrics before further optimization.

|              ns/cmp |               cmp/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|                4.35 |      229,642,637.11 |    0.2% |      1.10 | `CBlockIndexWorkComparator`

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|           38,213.19 |           26,168.97 |    0.5% |      1.10 | `CheckBlockIndex`

l0rinc committed:
refactor: optimize `arith_uint256` comparison with spaceship operator

Replace multiple comparisons with a single C++20 spaceship operator call directly.

|              ns/cmp |               cmp/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|                0.68 |    1,477,110,677.23 |    0.2% |      1.06 | `CBlockIndexWorkComparator`

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|           27,528.11 |           36,326.51 |    0.2% |      1.10 | `CheckBlockIndex`

l0rinc committed:
refactor: inline `arith_uint256` comparison operator

Remove the `CompareTo` method and inline its logic directly into `operator<=>`, updating related comments.
This eliminates function call overhead in the hot path during block generation and chain selection.

The comparison algorithm remains unchanged, iterating from most significant to least significant byte, but now returns `std::strong_ordering` directly, removing an extra conversion and enabling better inlining.

|              ns/cmp |               cmp/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|                2.18 |      459,159,880.91 |    0.2% |      1.10 | `CBlockIndexWorkComparator`

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|           31,780.27 |           31,466.07 |    0.5% |      1.10 | `CheckBlockIndex`

l0rinc committed:
fuzz: add equivalence coverage for comparison refactors

Add equivalence coverage to validate behavior while optimizing comparison operations.
The originals were kept exactly as the previous implementations were done.

The `arith_uint256` equivalence check is implemented as the `arith_uint256_comparison_equivalence` fuzz target.
It compares the refactored `operator<=>` against a snapshot of the old word-by-word comparison for all operators (<, >, <=, >=, ==, !=).

The `CBlockIndexWorkComparator` equivalence check is implemented as the `block_index_work_comparator` fuzz target.
It compares the refactored comparator against a snapshot of the old comparator logic (chain work, sequence ID, pointer tiebreak).

Add deterministic unit tests to lock in the expected ordering of both comparators.

Co-authored-by: Martin Zumsande <mzumsande@gmail.com>

Raimo33 commented:
Approach ACK

I have tested the new block index comparator but I’ll refrain from acking the added benchmarks/tests

l0rinc commented on src/bench/blockstorage.cpp:
```diff
@@ -0,0 +1,51 @@
+    }
+    std::ranges::shuffle(blocks, rng);
+
+    bench.run([&] {
```
Done, updated the commit messages and PR descriptions as well.

l0rinc commented:
Thanks for the measurements @Christewart, this is how your measurements compare to mine (but most importantly how it compares to `master`):
<img width="2385" height="883" alt="image" src="https://github.com/user-attachments/assets/96a59d25-292e-4e11-bd2c-3fb11a5b13cb" />

laanwj reviewed (approved):
Code review ACK c15d839aefba017e64f14cd9cd2779655352f4b6
i did not run the benchmarks but the code changes look good, using `<=>` makes sense.

mzumsande commented on src/test/arith_uint256_tests.cpp:
```diff
@@ -275,8 +275,43 @@ BOOST_AUTO_TEST_CASE( comparison ) // <= >= < >
+    BOOST_CHECK_LT(ZeroL, OneL);
+}
+
+BOOST_AUTO_TEST_CASE(comparison_equivalence)
```
commit msg 08d098947aaeae67e3aeef262df1ecdbdbed744a: wway ->way

mzumsande commented on src/test/blockchain_tests.cpp:
```diff
@@ -117,6 +117,43 @@ BOOST_AUTO_TEST_CASE(num_chain_tx_max)
     BOOST_CHECK_EQUAL(block_index.m_chain_tx_count, std::numeric_limits<uint64_t>::max());
 }
 
+BOOST_AUTO_TEST_CASE(cblockindex_comparator_equivalence)
```
this looks like a fuzz test, why not make it an actual fuzz target out of this?

mzumsande reviewed (commented):
Concept ACK

Not 100% sure yet about 2dd0f2ced35a268bcab661c671e7c70271cdd91f, seems like inlining gives most of the speedup, whereas the gain of using the spaceship operator (which comes at the cost of readability)  is marginal.

l0rinc commented on src/test/blockchain_tests.cpp:
```diff
@@ -117,6 +117,43 @@ BOOST_AUTO_TEST_CASE(num_chain_tx_max)
+        if (pa->nChainWork > pb->nChainWork) return false;
+        if (pa->nChainWork < pb->nChainWork) return true;
+
+        // ... then by earliest time received, ...
```
I have already done that, which parts do you think I'm missing? Edit: ah in the tests, good point, let me update it

l0rinc commented on src/node/blockstorage.h:
```diff
@@ -87,12 +87,20 @@ static constexpr uint32_t UNDO_DATA_DISK_OVERHEAD{STORAGE_HEADER_BYTES + uint256
 
 struct CBlockIndexWorkComparator {
-    bool operator()(const CBlockIndex* pa, const CBlockIndex* pb) const;
+    // First sort by most total work (descending), then by earliest activatable time (ascending), then by pointer value (ascending).
```
I wanted to use the SQL [order-by](https://www.postgresql.org/docs/current/queries-order.html#QUERIES-ORDER) terminology here
> Ascending order puts smaller values first, where “smaller” is defined in terms of the < operator.
Similarly, descending order is determined with the > operator.

But you're right, I inverted the terminology, added a dedicated test to check for hard-coded order (first param is increasing, second is decreasing), updated the comments (this is why I find comments a lazy solution, it's too easy to keep them up-to-date), added you as coauthor.

l0rinc commented:
Pushed, thanks for the review, you had a few good points, added you as coauthor.

> the gain of using the spaceship operator (which comes at the cost of readability) is marginal.

the spaceship might not be very familiar to us yet, but it's arguably simpler, having fewer moving parts, and some compilers prefer that over the manual comparisons.

l0rinc commented on src/test/blockchain_tests.cpp:
```diff
@@ -117,6 +120,79 @@ BOOST_AUTO_TEST_CASE(num_chain_tx_max)
+        if (pa->nSequenceId < pb->nSequenceId)
+            return false;
+        if (pa->nSequenceId > pb->nSequenceId)
+                return true;
```
done, thanks

l0rinc commented on src/node/blockstorage.h:
```diff
@@ -87,12 +87,21 @@ static constexpr uint32_t UNDO_DATA_DISK_OVERHEAD{STORAGE_HEADER_BYTES + uint256
+    // Pointer tiebreak should only happen with blocks loaded from disk, as those share the same id: 0 for blocks on the best chain, 1 for all others.
+    bool operator()(const CBlockIndex* pa, const CBlockIndex* pb) const noexcept
+    {
+        return std::tie(pa->nChainWork, pb->nSequenceId, pb)
```
I thought of eliminating the worst case here (when `pa == pb`, where it has to do the comparison for every byte) with something like:
```C++
return pa != pb &&
       std::tie(pa->nChainWork, pb->nSequenceId, pb)
     < std::tie(pb->nChainWork, pa->nSequenceId, pa);
```

But the benchmarks indicate that it's not really faster.

<details><summary>benchmark with pointer duplicates</summary>

```C++
// Copyright (c) 2025-present The Bitcoin Core developers
// Distributed under the MIT software license, see the accompanying
// file COPYING or http://www.opensource.org/licenses/mit-license.php.

#include <arith_uint256.h>
#include <bench/bench.h>
#include <chain.h>
#include <node/blockstorage.h>
#include <random.h>
#include <uint256.h>

#include <algorithm>
#include <memory>
#include <vector>

static void CBlockIndexWorkComparator(benchmark::Bench& bench)
{
    FastRandomContext rng{/*fDeterministic=*/true};

    constexpr size_t n{1'000};
    std::vector<std::shared_ptr<CBlockIndex>> blocks;
    blocks.reserve(n);
    for (size_t i{0}; i < n; ++i) {
        auto block{std::make_shared<CBlockIndex>()};
        block->nChainWork = UintToArith256(rng.rand256());
        block->nSequenceId = int32_t(rng.rand32());
        if (i % 10 == 1) {
            // Have some duplicates
            if (rng.randbool()) {
                block = blocks.back();
            } else {
                if (rng.randbool()) block->nChainWork = blocks.back()->nChainWork;
                if (rng.randbool()) block->nSequenceId = blocks.back()->nSequenceId;
            }
        }
        blocks.push_back(std::move(block));
    }
    std::ranges::shuffle(blocks, rng);

    const node::CBlockIndexWorkComparator comparator;
    bench.batch(n * n).unit("cmp").run([&] {
        for (size_t i{0}; i < n; ++i) {
            const auto* lhs{blocks[i].get()};
            for (size_t j{0}; j < n; ++j) {
                const auto* rhs{blocks[j].get()};
                const bool result{comparator(lhs, rhs)};
                ankerl::nanobench::doNotOptimizeAway(result);
            }
        }
    });
}

BENCHMARK(CBlockIndexWorkComparator, benchmark::PriorityLevel::HIGH);
```

</details>

b17504b611 refactor: optimize `CBlockIndexWorkComparator` with `std::tie`

|              ns/cmp |               cmp/s |    err% |         ins/cmp |         cyc/cmp |    IPC |        bra/cmp |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|                2.13 |      469,485,340.38 |    0.2% |           15.07 |            5.09 |  2.959 |           4.02 |    0.1% |     11.01 | `CBlockIndexWorkComparator`

|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|           90,265.10 |           11,078.48 |    0.1% |      637,536.00 |      215,946.18 |  2.952 |     178,844.00 |    0.0% |     10.84 | `CheckBlockIndex`


2b77e567d8 refactor: short-circuit `CBlockIndex` comparator for equality

|              ns/cmp |               cmp/s |    err% |         ins/cmp |         cyc/cmp |    IPC |        bra/cmp |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|                2.60 |      385,002,212.20 |    0.0% |           18.00 |            6.22 |  2.894 |           5.00 |    0.0% |     11.00 | `CBlockIndexWorkComparator`

|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|           94,110.53 |           10,625.80 |    0.0% |      682,454.03 |      225,242.05 |  3.030 |     186,486.01 |    0.0% |     11.00 | `CheckBlockIndex`

l0rinc commented:
> You should really add a fuzz test here to compare old and new behavior

Thanks for the hint, I've split the differential fuzzer and kept the simple unit tests o lock in the expected ordering of both comparators.
After a while (if fuzzers can't find any difference) we should be able to remove the fuzzers while keeping only the unit tests.

maflcko commented:
As pointed out earlier, most of the speedup comes from a move-only inline (https://github.com/bitcoin/bitcoin/pull/33637#pullrequestreview-3401558839)?

I think it would be better to just submit a trivial to review move-only change stand-alone, which wouldn't need in-depth review and dedicated differential unit/fuzz tests. Or at least the move-only change should be the first change. This way, it is easier to see if the other changes are worth it to review/test at all.

l0rinc committed:
move-only: inline `CBlockIndex` comparators to header

Move `CBlockIndexWorkComparator::operator()` and `CBlockIndexHeightOnlyComparator::operator()` to the header to facilitate inlining.

|              ns/cmp |               cmp/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|                2.87 |      348,283,873.30 |    0.2% |      1.10 | `CBlockIndexWorkComparator`

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|           36,534.37 |           27,371.49 |    0.3% |      1.10 | `CheckBlockIndex`

l0rinc committed:
refactor: optimize `CBlockIndexWorkComparator` with `std::tie`

Replace manual comparison branches with a single tuple comparison, allowing the compilers to generate more efficient comparison code.

Made them explicitly `noexcept`, which can enable small optimizer wins and aligns with expectations for hot-path comparators used in ordered containers.

|              ns/cmp |               cmp/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|                2.00 |      499,008,850.73 |    0.2% |      1.10 | `CBlockIndexWorkComparator`

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|           32,297.54 |           30,962.11 |    0.9% |      1.10 | `CheckBlockIndex`

Co-authored-by: Martin Zumsande <mzumsande@gmail.com>

l0rinc committed:
test: add sorting tests for `CBlockIndexWorkComparator` and `arith_uint256`

This test was added before other changes to create a baseline behavior.

optout21 commented:
reACK 703165e6d3b9fbd58b9c31ddce0e961047fc5e71
No significant new changes, no remarks.
P.S.: I'm aware this PR needs a rebase currently.

optout21 commented:
reACK 0d1119129fd2f9e7700702766121294b6fdf9d27
Rebased; no relevant changes, no remarks.
